### PR TITLE
Support Java package-level annotations

### LINKED ISSUE #156: Support package-level non-public marker
Instead of specifying a list of packages, it should be possible to specify a marker, for example [`@ApiStatus.Internal`](https://github.com/JetBrains/java-annotations/blob/b41de54f50e463211c71435c0fd86edeb9e3a152/common/src/main/java/org/jetbrains/annotations/ApiStatus.java#L73C25-L73C25)

### PR DESCRIPTION
**Pull request by fzhinkin**
Support extraction of packages annotated with particular annotations. These packages could be then be retained as public packages or filtered out as ignored packages.

Closes #156

### CODE REVIEW

File: src/main/kotlin/api/KotlinSignaturesLoading.kt [hunk 1]
```diff
@@ -218,6 +218,19 @@ private fun List<ClassBinarySignature>.filterOutNotAnnotated(
     }
 }
 
+@ExternalApi
+public fun List<ClassBinarySignature>.extractAnnotatedPackages(targetAnnotations: Set<String>): List<String> {
+    if (targetAnnotations.isEmpty()) return emptyList()
+
+    return filter {
+        it.name.endsWith("/package-info")
```
Comment (qwwdfsad) [inline src/main/kotlin/api/KotlinSignaturesLoading.kt:226]: This code looks like something in the form of "@Ann object \`package-info`" might slip in, but haven't checked it

File: src/main/kotlin/api/KotlinSignaturesLoading.kt [hunk 2]
```diff
@@ -218,6 +218,19 @@ private fun List<ClassBinarySignature>.filterOutNotAnnotated(
     }
 }
 
+@ExternalApi
```
Comment (qwwdfsad) [inline src/main/kotlin/api/KotlinSignaturesLoading.kt:221]: Could you please elaborate on why it's public API?

File: src/test/kotlin/tests/CasesPublicAPITest.kt [hunk 3]
```diff
@@ -59,11 +66,23 @@ class CasesPublicAPITest {
 
     private fun snapshotAPIAndCompare(testClassRelativePath: String, nonPublicMarkers: Set<String> = emptySet()) {
         val testClassPaths = baseClassPaths.map { it.resolve(testClassRelativePath) }
-        val testClasses = testClassPaths.flatMap { it.listFiles().orEmpty().asIterable() }
+        val testClasses = testClassPaths.flatMap {
+            if (!it.exists()) return@flatMap emptyList()
+            val allFiles = mutableListOf<File>()
+            Files.walkFileTree(it.toPath(), object : SimpleFileVisitor<Path>() {
```
Comment (qwwdfsad) [inline src/test/kotlin/tests/CasesPublicAPITest.kt:72]: We have a nice API for that:
```
it.toPath().visitFileTree { 
    onVisitFile { file, attributes ->
        allFiles.add(file.toFile())
    }
}
```

Also, worth filing KTIJ for the corresponding quickfix

File: src/main/kotlin/api/KotlinSignaturesLoading.kt (hunk 2 above)
Comment (fzhinkin) [reply src/main/kotlin/api/KotlinSignaturesLoading.kt:221]: It's an API facilitating markers gathering and if someone is using the BCV as an API, then they probably will need this as well.

Comment (qwwdfsad) [reply src/main/kotlin/api/KotlinSignaturesLoading.kt:221]: Maybe it's worth adding a bit of KDoc then

File: src/main/kotlin/api/KotlinSignaturesLoading.kt (hunk 1 above)
Comment (fzhinkin) [reply src/main/kotlin/api/KotlinSignaturesLoading.kt:234]: I narrowed the set of matching `package-info` declarations to what javac should actually emit (abstract synthetic interfaces), thanks for pointing to it!

**Review by qwwdfsad — APPROVED** [review]
Nice!

**Pull request merged**